test/images/stub-device-plugin/server.go

```diff
@@ -0,0 +1,316 @@
+)
+
+// StubDevicePlugin contains grpc info and devices info
+type StubDevicePlugin struct {
+	// grpc info
+	endpoint   string
+	grpcServer *grpc.Server
+
+	// devices of this DevicePlugin
+	mutex   sync.Mutex
+	devices map[string]pluginapi.Device
+
+	// dummy device file watcher
+	watcher     utilfs.FSWatcher
+	watcherPath string
+}
+
+func getDevices(dir string) map[string]pluginapi.Device {
+	files, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil
+	}
+
+	devices := make(map[string]pluginapi.Device)
+	reg := regexp.MustCompile(dummyDeviceRE)
+
+	// read from tmp dir
+	for _, f := range files {
+		if f.IsDir() {
+			continue
+		}
+
+		// match dummyDeviceRE
+		if reg.MatchString(f.Name()) {
+			log.Printf("found dummy device %q\n", f.Name())
+			devices[f.Name()] = pluginapi.Device{
+				ID: f.Name(),
+				// TODO(cph): read device state from dummy file
+				Health: pluginapi.Healthy,
+			}
+		}
+	}
+
+	log.Printf("getDevices: %+v\n", devices)
+	return devices
+}
+
+// NewStubDevicePlugin returns an initialized StubDevicePlugin.
+func NewStubDevicePlugin(endpoint, dir string) (*StubDevicePlugin, error) {
+	devices := getDevices(dir)
+	if len(devices) == 0 {
+		return nil, fmt.Errorf("No devices found in dir: %s", dir)
+	}
+
+	log.Printf("NewStubDevicePlugin, endpoint: %s, dir: %s", endpoint, dir)
+	return &StubDevicePlugin{
+		endpoint: endpoint + fmt.Sprintf("-%d", time.Now().Unix()),
+
+		devices: devices,
+
+		watcher:     utilfs.NewFsnotifyWatcher(),
+		watcherPath: dir,
+	}, nil
+}
+
+// Register registers the device plugin for the given resourceName with Kubelet.
+func (s *StubDevicePlugin) Register(kubeletEndpoint, resourceName string) error {
+	conn, err := grpc.Dial(kubeletEndpoint, grpc.WithInsecure(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}))
+	defer conn.Close()
+	if err != nil {
+		return err
+	}
+	client := pluginapi.NewRegistrationClient(conn)
+	reqt := &pluginapi.RegisterRequest{
+		Version:      pluginapi.Version,
+		Endpoint:     path.Base(s.endpoint),
+		ResourceName: resourceName,
+	}
+
+	_, err = client.Register(context.Background(), reqt)
+	if err != nil {
+		return fmt.Errorf("Cannot register stub worker to kubelet service: %v", err)
+	}
+	return nil
+}
+
+// ListAndWatch lists devices and update that list according to the Update call
+func (s *StubDevicePlugin) ListAndWatch(emtpy *pluginapi.Empty, server pluginapi.DevicePlugin_ListAndWatchServer) error {
+	for {
+		resp := new(pluginapi.ListAndWatchResponse)
+		for _, dev := range s.devices {
+			resp.Devices = append(resp.Devices, &pluginapi.Device{
+				ID:     dev.ID,
+				Health: dev.Health,
+			})
+		}
+		log.Printf("ListAndWatch: send devices %v", resp)
+
+		if err := server.Send(resp); err != nil {
+			log.Printf("device-plugin: cannot update device states: %v\n", err)
+			return err
+		}
+
+		time.Sleep(5 * time.Second)
```

Comment:
Couldn't we use a channel rather than a continuous loop ?
The ListAndWatch behavior is expected to be event driven rather than "interval-driven"

Reply:
@RenaudWasTaken Thanks so much, i add a channel here, PTAL :)